templates/common/_base/units/ovs-configuration.service.yaml

```diff
@@ -20,3 +20,5 @@ contents: |
 
   [Install]
   WantedBy=network-online.target
+  RequiredBy=crio.service
+  RequiredBy=kubelet.service
```

Comment:
This is in the install section, we already have on line 12, which will prevent kubelet or crio from starting until this service is complete.

Reply:
It seems to be a common pattern to include both - https://www.freedesktop.org/software/systemd/man/systemd.unit.html#%5BUnit%5D%20Section%20Options. Although I'm not sure why `Before=` is not sufficient here

Reply:
If `foo.service` is `After=bar.service`, but `bar.service` fails, then `foo.service` will still start.

`foo.service` must have `Requires=bar.service` in order for systemd to not start `foo` if `bar` fails.

Reply:
Right, but I thought the issue here is that you are seeing a race between ovs-configuration service and kubelet. Using After/Before should ensure that kubelet does not start until ovs-configuration is finished (regardless of if ovs-configuration succeeds or fails).

Are we sure the issue is a race between the two services, instead of an issue that ovs-configuration failed and then kubelet started anyway?

Reply:
Yes :( In [this OKD test](https://prow.ci.openshift.org/view/gs/origin-ci-test/logs/promote-release-openshift-okd-machine-os-content-e2e-aws-4.6/1296266834999250944) `ip-10-0-158-138` node started kubelet before ovs-configure has completed (see [worker logs](https://storage.googleapis.com/origin-ci-test/logs/promote-release-openshift-okd-machine-os-content-e2e-aws-4.6/1296266834999250944/artifacts/e2e-aws/nodes/workers-journal). That caused SDN container crashloop and failed the test.

Restarted OKD test twice and nodes come up as expected.

Reply:
@vrutkovs looking at your log I think I see you are concerned about node ip-10-0-158-138. I think the problem is the way the logs are munged. If you look at:
https://gist.github.com/trozet/232688e0ce7bd853cb720ee65b0cc357#file-ovs-config-kubelet-worker-race-L42

We see there kubelet is started without a message before it saying:
Aug 20 02:45:05.478151 ip-10-0-146-26 systemd[1]: ovs-configuration.service: Succeeded.

However, if you look at the log messages with no ip hostname before them:
https://gist.github.com/trozet/232688e0ce7bd853cb720ee65b0cc357#file-ovs-config-kubelet-worker-race-L46

You can see that ovs-configuration failed right before kubelet was started:
Aug 20 02:47:28.524450 localhost.localdomain systemd[1]: ovs-configuration.service: Failed with result 'exit-code'.
Aug 20 02:47:29.363963 ip-10-0-158-138 systemd[1]: Starting Kubernetes Kubelet...

Reply:
Oh, I see, so perhaps `ovs-configuration` should depend on `systemd-hostnamed` instead?

Reply:
Seems `After` is not sufficient, see [worker logs](https://storage.googleapis.com/origin-ci-test/pr-logs/pull/openshift_machine-config-operator/2013/pull-ci-openshift-machine-config-operator-master-okd-e2e-aws/1297192628235800576/artifacts/e2e-aws/gather-extra/nodes/workers-journal):
```
Aug 22 16:02:27.090063 localhost.localdomain systemd[1]: ovs-configuration.service: Main process exited, code=exited, status=1/FAILURE
Aug 22 16:02:27.090224 localhost.localdomain systemd[1]: ovs-configuration.service: Failed with result 'exit-code'.                                                                            Aug 22 16:02:27.091858 localhost.localdomain systemd[1]: Failed to start Configures OVS with proper host networking configuration.
....
Aug 22 16:02:27.127077 localhost.localdomain systemd[1]: Starting Container Runtime Interface for OCI (CRI-O)...
```
and the node fails. Re-added `RequiredBy` to install section

Reply:
I don't think the bug description is accurate. ovs-configuration is not racing with kubelet. Kubelet is simply running after it. Requires will ensure that kubelet only starts if ovs-configuration was successful. Since kubelet did not require network-online.target before I didn't want to make that a requirement here. I'm not convinced it should be a require.

ovs-configuration can race with setting the hostname. However, I don't see how that is a problem. ovs-configuration doesn't rely on the hostname being set. Alternatively, sometimes DHCP can offer a hostname. I'm not sure how the current hostname service works, but if it ever looks to use a hostname offered by DHCP then we should maybe hold off on the hostname service until ovs-configuration is done.

@vrutkovs were you able to determine why the ovs-configuration script was failing in the first place?

Reply:
> ovs-configuration is not racing with kubelet

Kubelet may start before ovs-configuration completes successfully. `RequiredBy` ensures that kubelet would wait for it complete.

>  ovs-configuration can race with setting the hostname

Correct, that seems to make it crash. But the real issue is that kubelet start regardless of the ovs-configuration exit status

Reply:
@vrutkovs in the logs you shared:
Aug 22 16:02:27.090224 localhost.localdomain systemd[1]: ovs-configuration.service: Failed with result 'exit-code'.
Aug 22 16:02:27.127077 localhost.localdomain systemd[1]: Starting Container Runtime Interface for OCI (CRI-O)...

CRIO is starting *after* ovs-configuration service. I'm still not seeing any issue here.

Reply:
Its starting after first run of ovs-configuration, but its doesn't wait for it to complete successfully.